src/ServiceControl.Monitoring/Infrastructure/EndpointRegistry.cs

```diff
@@ -6,18 +6,23 @@
 
     public class EndpointRegistry
     {
-        ConcurrentBag<EndpointInstanceId> endpointInstances = new ConcurrentBag<EndpointInstanceId>();
+        ConcurrentDictionary<EndpointInstanceId, bool> endpointInstances = new ConcurrentDictionary<EndpointInstanceId, bool>();
 
         public void Record(EndpointInstanceId endpointInstanceId)
         {
-            endpointInstances.Add(endpointInstanceId);    
+            endpointInstances.TryAdd(endpointInstanceId, true);    
         }
 
-        public IDictionary<string, IEnumerable<EndpointInstanceId>> AllEndpoints()
+        public IDictionary<string, IEnumerable<EndpointInstanceId>> GetAllEndpoints()
         {
-            return endpointInstances.ToArray()
+            return endpointInstances.Keys.ToArray()
```

Comment:
`endpointInstances.Select(kvp=>kvp.Keys) ` would not issue the lock (the non-locking concurrent iterator). Should we consider it in here?